### ISSUE TITLE
Download of specified time-series failed

### ISSUE THREAD

**Issue by cgarciga** [onnokleen/alfred#7]
Tried to run the following code from the vignette, and got the following error:

![image](https://user-images.githubusercontent.com/13739205/103016803-a41ba300-4510-11eb-88fd-ce75b5a49111.png)

See also #4

**Comment by onnokleen** [comment]
I still can't replicate the issue. Works on my system even though I did a clean install recently. Could you try reinstalling all dependencies?

**Comment by onnokleen** [comment]
Maybe also try to install curl:

install.packages('curl')

**Comment by aboslett** [comment]
I am observing the same issue. The pull will work intermittently. I have set up a loop that imports data across a number of variables from 1960 on, does some automatic seasonality adjustments, etc, 

The main code of interest is here:

  temp <- get_alfred_series(series_id = fff,
                            observation_start = '1960-01-01',
                            observation_end = lubridate::today()) 

![image](https://user-images.githubusercontent.com/33694626/105770770-a6b55200-5f2d-11eb-85f5-cb13f37eb452.png)

Yet, it worked moments later.

![image](https://user-images.githubusercontent.com/33694626/105770685-8c7b7400-5f2d-11eb-8330-eca6dc8f78c4.png)

**Comment by aboslett** [comment]
I should also note that the "curl" package is already installed and loaded.

**Comment by onnokleen** [comment]
I think the problem is the number of request that is currently allowed by the FRED API. I tried to ask for an increase but emails to webmaster@research.stlouisfed.org came back with a failure notice. I will try again at some point but I tried several email addresses of mine.

**Comment by Coop197823** [comment]
Looks like I am running into a similar issue this morning:

Using the example from the main page:

```
library(alfred)
# Download industrial production index releases from March 2015 for 2013.
get_alfred_series("INDPRO", "test",
                  observation_start = "2013-03-01", observation_end = "2013-03-30",
                  realtime_start = "2015-02-02", realtime_end = "2015-02-02")
# Wrapper for getting only most recent releases 
get_fred_series("INDPRO", "indpro", observation_start = "2009-03-01", observation_end = "2009-03-01")
```

I get error:

```
[1] "Download of specified time-series failed - did you misspell the identifier?"
[1] "Error in open.connection(con, \"rb\") : HTTP error 429.\n"
attr(,"class")
[1] "try-error"
attr(,"condition")
<simpleError in open.connection(con, "rb"): HTTP error 429.>
NULL
```

It was working fine yesterday

**Comment by SreyasMaha** [comment]
Was running into the same issue yesterday/today. My current fix is to redefine the get_alfred_series() function using my personal FRED API key. It could be helpful to add an 'api_key' argument to the functions!

**Comment by Coop197823** [comment]
Yep, It looks like the issue is this code uses the Creator's API key, but if we could use our own, then that might alleviate some of these issues.  Does anyone know what the limit on API calls there is in a set defined period that then causes this blockage?

**Comment by Coop197823** [comment]
> 
> 
> Was running into the same issue yesterday/today. My current fix is to redefine the get_alfred_series() function using my personal FRED API key. It could be helpful to add an 'api_key' argument to the functions!

Where are you adding in your api key?  Are you modifying the package files in the library?

**Comment by SreyasMaha** [comment]
> 
> 
> Yep, It looks like the issue is this code uses the Creator's API key, but if we could use our own, then that might alleviate some of these issues. Does anyone know what the limit on API calls there is in a set defined period that then causes this blockage?

It's a temporary solution but for the time being I just copy + pasted the source code from here: https://rdrr.io/cran/alfred/src/R/get_alfred_series.R and about halfway down in the paste sequence substituted my API key

**Comment by onnokleen** [comment]
Good idea, I will add this option

**Comment by Coop197823** [comment]
just as a quick question, as I work through the temporary workaround, I am getting an error when I run the code for the get_FRED function:

library(dplyr)
library(RJSONIO)

```
get_fred_series <- function(series_id, series_name = NULL,
                            observation_start = NULL, observation_end = NULL) {
  
  
  length_series_id <- nchar(series_id)
  
  if (!is.character(series_id)) {
    stop("series_id is always in characters")
  }
  
  if (is.null(series_name)) {
    series_name <- series_id
  }
  
  if (is.null(observation_start)) {
    observation_start <- "1776-07-04"
  }
  
  if (is.null(observation_end)) {
    observation_end <- "9999-12-31"
  }
  
  df_series <-
    try({
      fromJSON(
        paste0("https://api.stlouisfed.org/fred/series/observations?series_id=",
               series_id,
               "&observation_start=",
               observation_start,
               "&observation_end=",
               observation_end,
               "&output_type=2",
               "&api_key=MYAPI&file_type=json")
      )$observations
    }, silent = TRUE)
  
  if (class(df_series) == "try-error") {
    print("Download of specified time-series failed - did you misspell the identifier?")
    print(df_series)
    return(NULL)
  }
  
  df_series <-
    df_series %>%
    dplyr::mutate(date = as_date(.data$date))
  
  colnames(df_series)[!(colnames(df_series) %in% "date")] <- series_name
  df_series[, 2] <- as.numeric(unlist(df_series[, 2]))
  df_series
}
get_fred_series("INDPRO", "indpro", observation_start = "2009-03-01", observation_end = "2009-03-01")
```
and am getting the error:

```
 Error in UseMethod("mutate_") : 
  no applicable method for 'mutate_' applied to an object of class "list" 
```

Are you updating the df_series away from a list into a data frame?

**Comment by SreyasMaha** [comment]
> 
> 
> just as a quick question, as I work through the temporary workaround, I am getting an error when I run the code for the get_FRED function:
> 
> library(dplyr)
> library(RJSONIO)
> 
> ```
> get_fred_series <- function(series_id, series_name = NULL,
>                             observation_start = NULL, observation_end = NULL) {
>   
>   
>   length_series_id <- nchar(series_id)
>   
>   if (!is.character(series_id)) {
>     stop("series_id is always in characters")
>   }
>   
>   if (is.null(series_name)) {
>     series_name <- series_id
>   }
>   
>   if (is.null(observation_start)) {
>     observation_start <- "1776-07-04"
>   }
>   
>   if (is.null(observation_end)) {
>     observation_end <- "9999-12-31"
>   }
>   
>   df_series <-
>     try({
>       fromJSON(
>         paste0("https://api.stlouisfed.org/fred/series/observations?series_id=",
>                series_id,
>                "&observation_start=",
>                observation_start,
>                "&observation_end=",
>                observation_end,
>                "&output_type=2",
>                "&api_key=MYAPI&file_type=json")
>       )$observations
>     }, silent = TRUE)
>   
>   if (class(df_series) == "try-error") {
>     print("Download of specified time-series failed - did you misspell the identifier?")
>     print(df_series)
>     return(NULL)
>   }
>   
>   df_series <-
>     df_series %>%
>     dplyr::mutate(date = as_date(.data$date))
>   
>   colnames(df_series)[!(colnames(df_series) %in% "date")] <- series_name
>   df_series[, 2] <- as.numeric(unlist(df_series[, 2]))
>   df_series
> }
> get_fred_series("INDPRO", "indpro", observation_start = "2009-03-01", observation_end = "2009-03-01")
> ```
> 
> and am getting the error:
> 
> ```
>  Error in UseMethod("mutate_") : 
>   no applicable method for 'mutate_' applied to an object of class "list" 
> ```
> 
> Are you updating the df_series away from a list into a data frame?

Try calling library(jsonlite) instead of library(RJSONIO). I had a similar error when I used a package other than jsonlite for the fromJSON function (I can't remember which one it was, it might have been RJSONIO). If you're still getting an error I can post my packages but it looks like you made the change to the function in the right spot!

**Comment by Coop197823** [comment]
> Try calling library(jsonlite) instead of library(RJSONIO). I had a similar error when I used a package other than jsonlite for the fromJSON function (I can't remember which one it was, it might have been RJSONIO). If you're still getting an error I can post my packages but it looks like you made the change to the function in the right spot!

Thanks for the help!  so the jsonlite package helped resolve that error, now I am getting the error mentioned above from too many api requests.  I did request a new key from a new account, and again am still getting that too many request error.  Any thoughts as why that would be happening from every request?

**Comment by SreyasMaha** [comment]
> Thanks for the help! so the jsonlite package helped resolve that error, now I am getting the error mentioned above from too many api requests. I did request a new key from a new account, and again am still getting that too many request error. Any thoughts as why that would be happening from every request?

I tried the lines of example code you had initially posted and they are working on my end (with an unmodified get_fred_series() function as well, so I guess the API limit issues haven't been hit yet). I have had issues in the past due to proxy settings when on a VPN.

**Comment by onnokleen** [comment]
I am on my way to submit a new version that allows for supplying your own API key.

**Comment by onnokleen** [comment]
It's available on CRAN. I will close this issue for now.

Thanks for all the feedback!

**Comment by SreyasMaha** [comment]
The updated version is working perfectly, thank you for all of your work Onno!

**Comment by gillyparra** [comment]
Some variables still return "null"

"CURRDD","DEMDEPSL","M2SL","SVSTSL"-Variables I found that dont work.

Trying 
`get_alfred_series("CURRDD","currdd",)`
Or any of the others above returns null/nothing, I played around with the dates and it seems that there is a max length between the start and end dates. 
When trying for all dates, it does not give me an error either. 
Is there a way to get around this and pull all observations? Would I have to supply my own API key?
